src/librustc_typeck/outlives/implicit_infer.rs

```diff
@@ -0,0 +1,401 @@
+            //
+            // We also want to calculate potential predicates for the T
+            ty::TyRef(region, mt) => {
+                insert_outlives_predicate(tcx, mt.ty.into(), region, required_predicates);
+            }
+
+            // For each TyAdt (struct/enum/union) type `Foo<'a, T>`, we
+            // can load the current set of inferred and explicit
+            // predicates from `global_inferred_outlives` and filter the
+            // ones that are TypeOutlives.
+            //
+            ty::TyAdt(def, substs) => {
+                // First check the inferred predicates
+                //
+                // Example 1:
+                //
+                //     struct Foo<'a, T> {
+                //         field1: Bar<'a, T>
+                //     }
+                //
+                //     struct Bar<'b, U> {
+                //         field2: &'b U
+                //     }
+                //
+                // Here, when processing the type of `field1`, we would
+                // request the set of implicit predicates computed for `Bar`
+                // thus far. This will initially come back empty, but in next
+                // round we will get `U: 'b`. We then apply the substitution
+                // `['b => 'a, U => T]` and thus get the requirement that `T:
+                // 'a` holds for `Foo`.
+                if let Some(unsubstituted_predicates) = global_inferred_outlives.get(&def.did) {
+                    for unsubstituted_predicate in unsubstituted_predicates {
+                        // `unsubstituted_predicate` is `U: 'b` in the
+                        // example above.  So apply the substitution to
+                        // get `T: 'a` (or `predicate`):
+                        let predicate = unsubstituted_predicate.subst(tcx, substs);
+                        insert_outlives_predicate(
+                            tcx,
+                            predicate.0,
+                            predicate.1,
+                            required_predicates,
+                        );
+                    }
+                }
+
+                // We also have to check the explicit predicates
+                // declared on the type.
+                //
+                // Example 2:
+                //
+                //     struct Foo<'a, T> {
+                //         field1: Bar<T>
+                //     }
+                //
+                //     struct Bar<U> where U: 'static, U: Foo {
+                //         ...
+                //     }
+                //
+                // Here, we should fetch the explicit predicates, which
+                // will give us `U: 'static` and `U: Foo`. The latter we
+                // can ignore, but we will want to process `U: 'static`,
+                // applying the substitution as above.
+                if let Some(general_predicates) = explicit_map.get(&def.did) {
```

Comment:
@nikomatsakis `explicit_map` doesnt hold `RequiredPredicates` so and therefore the messier `match` arm. 

---
However I wonder if `RequiredPredicates` is necessary (although it would make the match arm nicer here) given your comment about smoothing the impedance matching in outlives/mod.rs

From what I understand, `RequiredPredicates` enforces a `Kind: Region` where Kind could be a `Type` or `Region`. That seems like a abstraction when we could simply enforce that all predicates added to the returning map be of the `TypeOutlives` or `RegionOutlives`. Additionally this might break down if `Kind` can later be something else also.